torch/utils/_pytree.py

```diff
@@ -1211,6 +1211,11 @@ def unflatten(self, leaves: Iterable[Any]) -> PyTree:
 
         return unflatten_fn(child_pytrees, self.context)
 
+    def __hash__(self) -> int:
```

Comment:
We cannot use the default hashing since list is not hashable. We need to hash the TreeSpec so that it could be put in to tensor_attrs map in _symbolic_trace.

Reply:
Could we implement this by dispatching the `self.type`?

For example, the `dict` tree spec is actually order-sensitive. It can be hashed via:

```python
if self.type is dict:
    return hash((self.type, tuple(self.context), tuple(self.children_specs)))
```

Same argument for `tuple`, `namedtuple`, `list`, `OrderedDict`, `defaultdict`, `deque`, and `torch.return_types.*`.

Reply:
hm...I feel this change is beyond the scope of this pr and probably need some more discussion and testing. Mind we add it as a TODO and implement it in a sperate PR?

Reply:
Upon a second thought, is it that you want to support the hashing of the listed types with `hash((self.type, tuple(self.context), tuple(self.children_specs)))` and raise notimplemented for other cases?